build.ps1

```diff
@@ -93,54 +84,40 @@ if($FoundDotNetCliVersion -ne $DotNetVersion) {
 $env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
 $env:DOTNET_CLI_TELEMETRY_OPTOUT=1
 
-###########################################################################
-# INSTALL NUGET
-###########################################################################
-
-# Make sure nuget.exe exists.
-$NugetPath = Join-Path $ToolPath "nuget.exe"
-if (Test-Path $NugetPath)
-{
-    [Version] $minVersion = [Version]'4.9.0.0'
-    [Version] $nugetVersion = Get-ChildItem $NugetPath | % VersionInfo| % FileVersion
-    if ($minVersion -gt $nugetVersion)
-    {
-        Write-Host "NuGet.exe version $nugetVersion to old deleting $NugetPath..."
-        Remove-Item $NugetPath
-    }
-}
-if (!(Test-Path $NugetPath)) {
-    Write-Host "Downloading NuGet.exe..."
-    (New-Object System.Net.WebClient).DownloadFile($NugetUrl, $NugetPath);
-}
 
 ###########################################################################
 # INSTALL CAKE
 ###########################################################################
 
 # Make sure Cake has been installed.
-$CakePath = Join-Path $ToolPath "Cake.$CakeVersion/Cake.exe"
-if (!(Test-Path $CakePath)) {
-    Write-Host "Installing Cake..."
-    Invoke-Expression "&`"$NugetPath`" install Cake -Version $CakeVersion -OutputDirectory `"$ToolPath`"" | Out-Null;
-    if ($LASTEXITCODE -ne 0) {
-        Throw "An error occurred while restoring Cake from NuGet."
-    }
+[string] $CakeExePath = ''
+[string] $CakeInstalledVersion = Get-Command dotnet-cake -ErrorAction SilentlyContinue  | % {&$_.Source --version}
+
+if ($CakeInstalledVersion -eq $CakeVersion) {
+    # Cake found locally
+    $CakeExePath = (Get-Command dotnet-cake).Source
+}
+else {
+    $CakePath = Join-Path $ToolPath ".store\cake.tool\$CakeVersion"
+    $CakeExePath = (Get-ChildItem -Path $ToolPath -Filter "dotnet-cake*" -File| ForEach-Object FullName | Select-Object -First 1)
+
+    if ((!(Test-Path -Path $CakePath -PathType Container)) -or (!(Test-Path $CakeExePath -PathType Leaf))) {
+        & dotnet tool install --tool-path $ToolPath --version $CakeVersion Cake.Tool
+        if ($LASTEXITCODE -ne 0)
+        {
+            'Failed to install cake'
+            exit 1
+        }
+        $CakeExePath = (Get-ChildItem -Path $ToolPath -Filter "dotnet-cake*" -File| ForEach-Object FullName | Select-Object -First 1)
+    }   
 }
 
 ###########################################################################
 # RUN BUILD SCRIPT
 ###########################################################################
-
-# Build the argument list.
-$Arguments = @{
-    target=$Target;
-    configuration=$Configuration;
-    verbosity=$Verbosity;
-    dryrun=$WhatIf;
-}.GetEnumerator() | %{"--{0}=`"{1}`"" -f $_.key, $_.value };
-
-# Start Cake
-Write-Host "Running build script..."
-Invoke-Expression "& `"$CakePath`" `"build.cake`" $Arguments $ScriptArgs"
-exit $LASTEXITCODE
+& "$CakeExePath" ./build.cake --bootstrap
+if ($LASTEXITCODE -eq 0)
+{
+    & "$CakeExePath" ./build.cake $args
```

Comment:
I think for this to work as intended you need to [splat](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_splatting?view=powershell-5.1&viewFallbackFrom=powershell-Microsoft.PowerShell.Core) the args variable.

```suggestion
    & "$CakeExePath" ./build.cake @args
```

Reply:
Not at computer now, so can't test, but pretty sure `$args` works with native executables as all are passed unparsed.

Won't splat try match to match arguments against named PowerShell cmd:let arguments?

Reply:
I absolutely do not understand the finicky inner workings of PowerShell, but at least in my extremely similar bootstrapping script I'm working on passing it as `$args` did not work and it was like no arguments were passed at all. For example doing something like `./build.ps1 --verbosity=verbose` did not result in verbose output but did when I switched to it use `@args` instead.

Reply:
Ok, I'll give it a test run when back at computer

Looking at CI builds

`.\build.ps1 --target="AppVeyor"` seems to have worked at least.

Reply:
It was actually a bad example that I gave. It seems to work with exactly one parameter. I was using multiple and the first one was `-verbose` in an attempt to make the bootstrapping script also run verbosely. That resulted in no errors but no parameters effecting the cake script. When I take that out it works with a single `--parameter=value` type parameter but fails when using multiple. So a better example to try is this something like this:

```shell
./build.ps1 --verbosity=verbose --target=clean
```

That will result in the error:

> The value 'verbose --target=clean' is not a valid verbosity.

So it seems like it might be doing some kind of automatic quoting and passing them all as a single parameter (which is probably why it works with only one parameter). Maybe it's a subtle difference in behavior from removing the use of `Invoke-Expression`?

Reply:
Interesting both $args and @args work for me with multiple args, when I tried

```
cmd /c echo @args
cmd /c echo $args
```

Called `.\build.ps1 --target=Clean --verbosity=Diagnostic --test="jjkkjh    ddd"`

Output was very similar
```
--target=Clean --verbosity=Diagnostic "--test=jjkkjh    ddd"
--target=Clean --verbosity=Diagnostic "--test=jjkkjh    ddd"
```

Interesting though that both move the quotes, but it still work in the cake command line parser.

Maybe you had extra quotes?

Reply:
I looked at it again and figured out what the real difference was between our scripts. I was calling cake from within a PowerShell function and passing the script's Args into the function and then on to the `&` call operator. Something in that messed up the arguments and splatting just happened to fix them.

Sorry for the distraction. I was writing a bootstrapper and trying to base it on the latest and greatest version from this PR and hit an issue with the parameters that took me a while to get working and was just trying to share what I learned... I just didn't actually learn what I thought I learned and wasted your time instead. 🤦‍♂️

Reply:
No worries 😎